File: springboot/src/main/java/com/boilerplatecombo/controller/PostgresResearcherController.java
package com.boilerplatecombo.controller;

import com.boilerplatecombo.entity.postgres.PostgresResearcher;
import com.boilerplatecombo.entity.postgres.PostgresResearcherPage;
import com.boilerplatecombo.service.PostgresResearcherService;
import org.springframework.context.annotation.Lazy;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.nio.charset.StandardCharsets;

@RestController
@RequestMapping("/postgres/researcher")
public class PostgresResearcherController {

    private PostgresResearcherService service;

    public PostgresResearcherController(@Lazy PostgresResearcherService researcherService) {
        this.service = researcherService;
    }

    @GetMapping
    public ResponseEntity<PostgresResearcherPage> getResearchers(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "25") int pageSize,
            @RequestParam(defaultValue = "id") String sortBy,
            @RequestParam(defaultValue = "true") boolean ascending,
            @RequestParam(required = false) String filter
    ) {
        return ResponseEntity.ok(service.getResearchers(page, pageSize, sortBy, ascending, filter));
    }

    @GetMapping("/export")
    public ResponseEntity<byte[]> exportResearchers() {
        byte[] csv = service.exportResearchersAsCsv().getBytes(StandardCharsets.UTF_8);
        return ResponseEntity.ok()
                .header("Content-Disposition", "attachment; filename=researchers.csv")
                .header("Content-Type", "text/csv")
                .body(csv);
    }

    @PostMapping
    public ResponseEntity<PostgresResearcher> addResearcher(@RequestBody PostgresResearcher researcher) {
        PostgresResearcher added = service.addResearcher(researcher);
        return ResponseEntity.status(201).body(added);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Void> updateResearcher(@PathVariable Long id, @RequestBody PostgresResearcher researcher) {
        boolean updated = service.updateResearcher(id, researcher);
        return updated ? ResponseEntity.noContent().build() : ResponseEntity.notFound().build();
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteResearcher(@PathVariable Long id) {
        boolean deleted = service.deleteResearcher(id);
        return deleted ? ResponseEntity.noContent().build() : ResponseEntity.notFound().build();
    }
}

File: dotnet/Controllers/PostgresResearcherController.cs
﻿using BoilerplateCombo.Models;
using BoilerplateCombo.Service;
using Microsoft.AspNetCore.Mvc;

namespace BoilerplateCombo.Controllers;

[ApiController]
[Route("postgres/researcher")]
public class PostgresResearcherController : ControllerBase
{
    private readonly PostgresResearcherService _service;

    public PostgresResearcherController(PostgresResearcherService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<IActionResult> GetResearchers(int page = 1, int pageSize = 25, string sortBy = "id", bool ascending = true, string? filter = null)
    {
        var (researchers, totalCount) = await _service.GetResearchersAsync(page, pageSize, sortBy, ascending, filter);
        return Ok(new { data = researchers, totalCount });
    }

    [HttpGet("export")]
    public async Task<IActionResult> ExportResearchers()
    {
        var csv = await _service.ExportResearchersAsCsvAsync();
        return File(System.Text.Encoding.UTF8.GetBytes(csv), "text/csv", "researchers.csv");
    }

    [HttpPost]
    public async Task<IActionResult> AddResearcher([FromBody] Researcher researcher)
    {
        var added = await _service.AddResearcherAsync(researcher);
        return CreatedAtAction(nameof(GetResearchers), new { id = added.id }, added);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateResearcher(int id, [FromBody] Researcher researcher)
    {
        var updated = await _service.UpdateResearcherAsync(id, researcher);
        if (!updated)
            return NotFound();
        return NoContent();
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteResearcher(int id)
    {
        var deleted = await _service.DeleteResearcherAsync(id);
        if (!deleted)
            return NotFound();
        return NoContent();
    }
}